File: src/types/tracking.ts

// types/tracking.ts
export interface TrackingConfig {
  tracker_type: 'centroid' | 'kalman' | 'deep_sort' | 'byte_track';
  tracker_params: {
    max_disappeared?: number;
    max_distance?: number;
    use_kalman?: boolean;
  };
  speed_config?: {
    fps: number;
    pixel_to_meter_ratio: number;
  };
}

export interface ModelConfig {
  name: string;
  classFilter?: string[];
}

export interface DetectionResult {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
  confidence: number;
  class_id: number;
  label: string;
}

export interface TrackedObjectData {
  track_id: string;
  class_name: string;
  class_id: number;
  bbox: [number, number, number, number];
  centroid: [number, number];
  confidence: number;
  age: number;
  hits: number;
  time_since_update: number;
  velocity: [number, number];
  speed_info?: {
    speed_px_per_sec: number;
    speed_m_per_sec: number;
    direction: number;
    avg_speed_px_per_sec: number;
    avg_speed_m_per_sec: number;
  };
  trajectory_length: number;
}

export interface TrackingResults {
  tracked_objects: Record<string, TrackedObjectData>;
  zone_occupancy: Record<string, string[]>;
  speed_analysis: Record<string, any>;
  summary: {
    total_tracks: number;
    active_tracks: number;
    class_counts: Record<string, number>;
  };
}

File: app/schemas/tracking.py
from pydantic import BaseModel
from typing import Dict, Tuple, List, Optional, Any


class TrackedObject(BaseModel):
    track_id: str
    class_name: str
    class_id: int
    bbox: Tuple[float, float, float, float]
    centroid: Tuple[float, float]
    confidence: float
    age: int
    velocity: Tuple[float, float]
    speed_px_per_sec: float = 0.0
    speed_m_per_sec: Optional[float] = None
    distance_traveled: float = 0.0


class TrackingResponse(BaseModel):
    camera_id: str
    timestamp: int
    tracked_objects: Dict[str, TrackedObject]
    summary: Dict[str, Any]
    type: str = "tracking"


class ZoneDefinition(BaseModel):
    zone_id: str
    polygon_points: List[Tuple[int, int]]


class TrackingConfig(BaseModel):
    max_disappeared: int = 30
    max_distance: float = 100.0
    enable_speed: bool = True
    enable_zones: bool = False